### PR TITLE
Add observation delays for sensor latency modeling

### PR DESCRIPTION
**Pull request by kevinzakka**
Implements a delay buffer and integrates it with `ObservationManager`. Policies can train with delayed (stale) observations to model sensor latency, improving sim-to-real transfer for robots with communication delays or slow sensor refresh rates.

## Changes

**DelayBuffer**
- Wraps `CircularBuffer` to serve delayed frames
- Samples lag uniformly from `[min_lag, max_lag]`, either per-env or shared
- Supports "hold" probability to reuse previous lag (temporal correlation)
- Multi-rate: refresh lags every `update_period` steps with optional phase offsets

**ObservationManager**
- Adds 6 term-level delay parameters: `delay_min_lag`, `delay_max_lag`, `delay_per_env`, `delay_hold_prob`, `delay_update_period`, `delay_per_env_phase`
- Pipeline: `compute → noise → clip → scale → delay → history`

## Usage

**Camera with 40-60ms latency at 50Hz control**
```python
@dataclass
class PolicyCfg(ObservationGroupCfg):
  # 40-60ms latency = 2-3 steps at 50Hz (20ms/step)
  camera: ObservationTermCfg = ObservationTermCfg(
    func=camera_obs,
    delay_min_lag=2,  # 40ms
    delay_max_lag=3,  # 60ms
  )
```

**Multi-rate sensor system (50Hz control)**
```python
@dataclass
class PolicyCfg(ObservationGroupCfg):
  # Fast encoder: no inherent delay
  joint_pos: ObservationTermCfg = ObservationTermCfg(func=joint_pos)

  # 30Hz camera: 2-step refresh period + 40-60ms latency
  camera: ObservationTermCfg = ObservationTermCfg(
    func=camera_obs,
    delay_min_lag=2,          # 40ms
    delay_max_lag=3,          # 60ms
    delay_update_period=2,    # ~30Hz refresh (50Hz / 2)
    delay_per_env_phase=True, # Stagger updates across envs
  )
```

**Combined with history (delayed frame stacking):**
```python
@dataclass
class PolicyCfg(ObservationGroupCfg):
  joint_vel: ObservatioqnTermCfg = ObservationTermCfg(
    func=joint_vel,
    delay_min_lag=1,
    delay_max_lag=3,          # 20-60ms variable delay
    history_length=5,         # Stack 5 delayed frames
    flatten_history_dim=True,
  )
  # Pipeline: compute → delay → stack → flatten
```

## Notes
- Delays are quantized to control steps
- Convert to ms: `lag_steps * (1000 / control_hz)`
- Use `min_lag=max_lag` for constant delay; use range for stochastic variation
- Set `per_env=False` to share lag across all envs

### CODE REVIEW

File: docs/api/observation_history_delay.md [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+buffer = [obs_0, obs_1, obs_2]  # Normal accumulation
+```
+
+## Observation Delay
+
+Real robots have sensors with communication delays (WiFi, USB) and varying refresh
+rates (30Hz camera, 100Hz encoders). The delay system models both sensor latency
+and slower-than-control refresh rates.
+
+### Delay Parameters
```
Comment (brentyi) [inline docs/api/observation_history_delay.md:113]: Seems like it would be easy for this docs organization to get stale with future code updates; in the future we could explore co-locating the docs for a specific API with the implementation + maybe Sphinx for generating docs?

File: src/mjlab/managers/manager_term_config.py [hunk 2]
```diff
@@ -86,18 +86,41 @@ class EventTermCfg(ManagerTermBaseCfg):
...
 
   noise: NoiseCfg | NoiseModelCfg | None = None
   """Noise model to apply to the observation. Defaults to None."""
   clip: tuple[float, float] | None = None
   """Range (min, max) to clip the observation values. Defaults to None."""
   scale: tuple[float, ...] | float | torch.Tensor | None = None
   """Scaling factor(s) to multiply the observation by. Defaults to None."""
+  delay_min_lag: int = 0
+  """Minimum lag (in steps) for delayed observations. Lag sampled uniformly from
+  [min_lag, max_lag]. Convert to ms: lag * (1000 / control_hz). Default 0 (no delay)."""
```
Comment (brentyi) [inline src/mjlab/managers/manager_term_config.py]: also note for future: the default in docstring seems redundant

File: src/mjlab/utils/buffers/__init__.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 """Buffer utilities."""
 
 from mjlab.utils.buffers.circular_buffer import CircularBuffer
+from mjlab.utils.buffers.delay_buffer import DelayBuffer
 
-__all__ = ["CircularBuffer"]
+__all__ = ["CircularBuffer", "DelayBuffer"]
```
Comment (brentyi) [inline src/mjlab/utils/buffers/__init__.py]: also note for future: we could drop `__all__` and replace with `import X as X` convention for exports?

File: src/mjlab/utils/buffers/delay_buffer.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+    Returns:
+      New lags with shape (batch_size,).
+    """
+    new_lags = self._current_lags.clone()
+
+    if not torch.any(mask):
+      return new_lags
+
+    num_to_sample = int(mask.sum().item())
```
Comment (brentyi) [inline src/mjlab/utils/buffers/delay_buffer.py]: I'm nervous about the (`if not torch.any(...)`), calls to `.item()`, etc in this method. It looks like it would introduce a lot of syncs into the CUDA stream. but maybe overhead is negligible?

File: src/mjlab/utils/buffers/delay_buffer.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+      dimension of inputs). Defaults to 1.
+    device (str, optional): Torch device for storage and RNG (e.g., `"cpu"`,
+      `"cuda"`). Defaults to `"cpu"`.
+    per_env (bool, optional): If True, sample a separate lag per environment;
+      otherwise sample one lag and share it across environments. Defaults to True.
+    hold_prob (float, optional): Probability in `[0.0, 1.0]` to keep the previous
+      lag when an update would occur. Defaults to 0.0.
+    update_period (int, optional): If > 0, refresh lags every `N` steps per
+      environment; if 0, consider updating every step. Defaults to 0.
+    per_env_phase (bool, optional): If True and `update_period > 0`, each
```
Comment (brentyi) [inline src/mjlab/utils/buffers/delay_buffer.py:92]: nit: `generator` is missing from this docstring

File: src/mjlab/utils/buffers/delay_buffer.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+  @property
+  def is_initialized(self) -> bool:
+    """Check if buffer has been initialized with at least one append."""
+    return self._buffer.is_initialized
+
+  @property
+  def current_lags(self) -> torch.Tensor:
+    """Current lag per environment. Shape: (batch_size,)."""
+    return self._current_lags.clone()
```
Comment (brentyi) [inline src/mjlab/utils/buffers/delay_buffer.py]: is this clone necessary?

**Review by brentyi — APPROVED** [review]
:+1: left some small comments but lgtm!

I guess it'll take some effort to resolve merge conflicts in #124 :grimacing:

File: docs/api/observation_history_delay.md (hunk 1 above)
Comment (kevinzakka) [reply docs/api/observation_history_delay.md:113]: Indeed, will do when we are close to stable release!

File: src/mjlab/managers/manager_term_config.py (hunk 2 above)
Comment (kevinzakka) [reply src/mjlab/managers/manager_term_config.py]: Removed them.

File: src/mjlab/utils/buffers/__init__.py (hunk 3 above)
Comment (kevinzakka) [reply src/mjlab/utils/buffers/__init__.py]: Done.

File: src/mjlab/utils/buffers/delay_buffer.py (hunk 4 above)
Comment (kevinzakka) [reply src/mjlab/utils/buffers/delay_buffer.py]: This should now be fixed!

File: src/mjlab/utils/buffers/delay_buffer.py (hunk 5 above)
Comment (kevinzakka) [reply src/mjlab/utils/buffers/delay_buffer.py:92]: Done.

File: src/mjlab/utils/buffers/delay_buffer.py (hunk 6 above)
Comment (kevinzakka) [reply src/mjlab/utils/buffers/delay_buffer.py]: Removed.

File: src/mjlab/utils/buffers/delay_buffer.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+      )
+      candidate_lags = shared_lag.expand(self.batch_size)
+
+    if self.hold_prob > 0.0:
+      should_sample = (
+        torch.rand(self.batch_size, device=self.device, generator=self.generator)
+        >= self.hold_prob
+      )
+    else:
+      should_sample = torch.ones(self.batch_size, dtype=torch.bool, device=self.device)
```
Comment (brentyi) [inline src/mjlab/utils/buffers/delay_buffer.py]: could avoid this extra tensor by just setting `update_mask = mask` here, and only applying binary `&` in the condition above

Comment (kevinzakka) [reply src/mjlab/utils/buffers/delay_buffer.py]: Fixed, good catch!
